### PR TITLE
Refactor ContractCallEvmCodesTest

### LINKED ISSUE #8450: Refactor `ContractCallEvmCodesTest`

### PR DESCRIPTION
**Pull request by kselveliev**
**Description**:

This PR refactors **ContractCallEvmCodesTest**.
**ContractCallEvmCodes** now is independent from **ContractCallTestSetup** and is independent of persistEntities method.
Extracted historical tests -> they are working as before as currently we are able to integrate only latest block tests with the web3j plugin.
Extracted tests that use another contract **TestAddressThis** into another class since it is unrelated to evm codes and will be refactored in a later PR.

 

This PR modifies ... in order to support ...
* Add ContractCallAddressThis - moved tests using the address this into this class
* Modified ContractCallEvmCodesTest - refactored the test using the web3j plugin and make independent
* Added ContractCallEvmCodesTestHistorical - use old version of tests for historical 

**Related issue(s)**:

Fixes #8450

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java [hunk 1]
```diff
@@ -16,347 +16,300 @@
...
-                "getBlockHash", EVM_CODES_ABI_PATH, recordFileForBlockHash.getIndex());
-        final var serviceParameters = serviceParametersForEvmCodes(functionHash);
-
-        assertThat(contractCallService.processCall(serviceParameters))
-                .isEqualTo("0x" + recordFileForBlockHash.getHash().substring(0, 64));
+    void getBlockHashReturnsCorrectHash() throws Exception {
+        final var contract = testWeb3jService.deploy(EvmCodes::deploy);
+        final var recordFileForBlockHash = domainBuilder.recordFile().persist();
+        // We need to wait enough time for the record file to be processed
+        Thread.sleep(500);
```
Comment (kselveliev) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:161]: Try to find a way to improve that.

**Comment by kselveliev** [comment]
Need to figure out why codeHash test fails

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        final var expectedGasUsed = gasUsedAfterExecution(serviceParameters);
+
+        assertThat(isWithinExpectedGasRange(
+                        longValueOf.applyAsLong(contractCallService.processCall(serviceParameters)), expectedGasUsed))
+                .isTrue();
+    }
+
+    @Test
+    void addressThisEthCallWithoutEvmAlias() {
+        String addressThisContractAddressWithout0x =
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java:69]: Nit, String can be var.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                ADDRESS_THIS_CONTRACT_ADDRESS.toString().substring(2);
+        String successfulResponse = "0x000000000000000000000000" + addressThisContractAddressWithout0x;
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java:71]:
```suggestion
        var successfulResponse = "0x" + StringUtils.leftPad( addressThisContractAddressWithout0x, 64, '0');
```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                                .toArrayUnsafe())
+                        .value(Bytes.fromHexString("0x4746573740000000000000000000000000000000000000000000000000000000")
+                                .toArrayUnsafe()))
+                .persist();
+
+        domainBuilder
+                .recordFile()
+                .customize(f -> f.bytes(addressThisContractBytes))
+                .persist();
+        return addressThisContractEntityId;
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java:138]: Return value not used.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java (hunk 2 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java]: Comment addressed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java (hunk 4 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java]: Commend addressed, removed return value and made method void.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java (hunk 3 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java]: Comment addressed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java [hunk 5]
```diff
@@ -16,347 +16,299 @@
...
-                .isEqualTo(
-                        "0xba80a53f981c4d0d6a2797b69f12f6e94c212f14685ac4b74b12bb6fdbffa2d17d87c5392aab792dc252d5de4533cc9518d38aa8dbf1925ab92386edd4009923");
+    void BLAKE2BFPrecompileContract() throws Exception {
+        final var contract = testWeb3jService.deploy(EvmCodes::deploy);
+        var result = contract.call_blake2().send();
+        String expectedResultHexString =
+                "0xba80a53f981c4d0d6a2797b69f12f6e94c212f14685ac4b74b12bb6fdbffa2d17d87c5392aab792dc252d5de4533cc9518d38aa8dbf1925ab92386edd4009923";
+        List<byte[]> expetedList = convertBlake2HexToList(expectedResultHexString.substring(2));
+        assertThat(result.get(0)).isEqualTo(expetedList.get(0));
+        assertThat(result.get(1)).isEqualTo(expetedList.get(1));
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:145]: Something like this should simplify things, and then you can remove `convertBlake2HexToList`: 

```suggestion
final var contract = testWeb3jService.deploy(EvmCodes::deploy);
        List<byte[]> result = contract.call_blake2().send();
        var expectedResultHexString =
                "ba80a53f981c4d0d6a2797b69f12f6e94c212f14685ac4b74b12bb6fdbffa2d17d87c5392aab792dc252d5de4533cc9518d38aa8dbf1925ab92386edd4009923";
        var resultBytes = Bytes.concatenate(result.stream().map(Bytes::wrap).toArray(Bytes[]::new)).toArray();
        var expected = hexStringToByteArray(expectedResultHexString);
        assertArrayEquals(resultBytes, expected);
```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java [hunk 6]
```diff
@@ -16,347 +16,299 @@
...
-    void getLatestBlockHashIsNotEmpty() {
-        final var functionHash = functionEncodeDecoder.functionHashFor("getLatestBlockHash", EVM_CODES_ABI_PATH);
-        final var serviceParameters = serviceParametersForEvmCodes(functionHash);
-
-        assertThat(contractCallService.processCall(serviceParameters))
-                .isNotEqualTo("0x0000000000000000000000000000000000000000000000000000000000000000");
+    void getLatestBlockHashIsNotEmpty() throws Exception {
+        final var contract = testWeb3jService.deploy(EvmCodes::deploy);
+        var result = contract.call_getLatestBlockHash().send();
+        var expectedResult = hexStringToByteArray("0000000000000000000000000000000000000000000000000000000000000000");
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:183]: Can create a constant for `0000000000000000000000000000000000000000000000000000000000000000`, then replace the six usages.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java [hunk 7]
```diff
@@ -16,347 +16,299 @@
...
-        final var expectedGasUsed = gasUsedAfterExecution(serviceParameters);
-
-        assertThat(isWithinExpectedGasRange(
-                        longValueOf.applyAsLong(contractCallService.processCall(serviceParameters)), expectedGasUsed))
-                .isTrue();
+    void testSystemContractCodeHash(String input, String expectedOutput) throws Exception {
+        final var contract = testWeb3jService.deploy(EvmCodes::deploy);
+        var result = contract.call_getCodeHash(input).send();
+        var expectedResult = hexStringToByteArray(expectedOutput);
+        assertThat(result).isEqualTo(expectedResult);
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:208]: Looks like the expectedOutput is always the same, can reference constant value directly and not pass in the second parameter.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java [hunk 8]
```diff
@@ -16,347 +16,299 @@
...
-                .value(0L)
-                .receiver(Address.ZERO)
-                .callData(callData)
-                .callType(ETH_ESTIMATE_GAS)
-                .block(BlockType.LATEST)
-                .gas(15_000_000L)
-                .isStatic(false)
-                .isEstimate(true)
-                .build();
+        return data;
```
Comment (edwin-greene) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:312]: It seems like there should be a library to do this for us; however I tried a few and all tests pass except `testNonSystemContractEthCallCodeHash()`. I am unsure why that test fails with standard hex to byte[] conversions.

File: hedera-mirror-web3/src/test/solidity/EvmCodes.sol [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    function getCodeHash(address _address) external view returns (bytes32) {
+        bytes32 codehash;
+        assembly {
+            codehash := extcodehash(_address)
+        }
+        return codehash;
+    }
+
+    function destroyContract(address payable beneficiary) public {
+        selfdestruct(beneficiary);
```
Comment (jascks) [inline hedera-mirror-web3/src/test/solidity/EvmCodes.sol:234]: While running `ContractCallAddressThisTest`, there are a number of Solidity compilation warnings reported.  Are these okay/expected? In this file, there is just the one below, but in `EthCall.sol`, there are five more.

```
Warning: "selfdestruct" has been deprecated. Note that, starting from the Cancun hard fork, the underlying opcode no longer deletes the code and data associated with an account and only transfers its Ether to the beneficiary, unless executed in the same transaction in which the contract was created (see EIP-6780). Any use in newly deployed contracts is strongly discouraged even if the new behavior is taken into account. Future changes to the EVM might further reduce the functionality of the opcode.
   --> src/test/solidity/EvmCodes.sol:234:9:
    |
234 |         selfdestruct(beneficiary);
    |         ^^^^^^^^^^^^```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    void testSystemContractCodeHashPreVersion38(String input) {
+        final var serviceParameters = serviceParametersForExecution(
+                Bytes.fromHexString(input),
+                EVM_CODES_CONTRACT_ADDRESS,
+                ETH_CALL,
+                0L,
+                BlockType.of(String.valueOf(EVM_V_34_BLOCK)));
+
+        assertThatThrownBy(() -> contractCallService.processCall(serviceParameters))
+                .isInstanceOf(MirrorEvmTransactionException.class);
```
Comment (jascks) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java:56]: Should dig deeper. I can see in the debugger that the exception message is "INVALID_SOLIDITY_ADDRESS" which is indicated in the comment above. So, should assert the message is as expected not just the exception type.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java (hunk 5 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java]: Comment addressed. I applied the above suggestion.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java (hunk 6 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java]: Comment addressed. I created a constant.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java (hunk 7 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:222]: Comment addressed.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java (hunk 8 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java]: I fixed the failing test. Now using ByteString.fromHex().toByteArray() allowed removing the custom method.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java (hunk 10 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java]: You are right. Comment addressed. I am now also asserting for the expected exception message.

File: hedera-mirror-web3/src/test/solidity/EvmCodes.sol (hunk 9 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/solidity/EvmCodes.sol:234]: Those are okay. I've actually refactored the EthCall.sol which should stop the warnings from appearing. I think we should still not change the selfdestruct one which will leave only one warning in the ci.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java (hunk 10 above)
Comment (jascks) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java]: Looks good. Thank you!

File: hedera-mirror-web3/src/test/solidity/EvmCodes.sol (hunk 9 above)
Comment (jascks) [reply hedera-mirror-web3/src/test/solidity/EvmCodes.sol:234]: Much nicer. 😄

**Review by jascks — APPROVED** [review]
LGTM

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+import com.hedera.mirror.web3.service.model.ContractExecutionParameters;
+import com.hedera.mirror.web3.viewmodel.BlockType;
+import com.hedera.node.app.service.evm.store.models.HederaEvmAccount;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.testcontainers.shaded.org.apache.commons.lang3.StringUtils;
+
+public class ContractCallAddressThisTest extends ContractCallTestSetup {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java:35]:
```suggestion
class ContractCallAddressThisTest extends ContractCallTestSetup {
```

Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java:35]: As I understand it, this is just moving things and we will refactor later to not extend `ContractCallTestSetup`. Can you please open a ticket under the epic for this future work?

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.tuweni.bytes.Bytes;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+public class ContractCallEvmCodesHistoricalTest extends ContractCallTestSetup {
+
+    @BeforeEach
+    void beforeAll() {
+        evmCodesContractPersist();
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java:35]: Even if we can't use web3j to call the historical contract, we need to figure out a way to not depend upon our manually generated `.bin` and `.json` files. Can be a separate ticket.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java [hunk 13]
```diff
@@ -16,347 +16,309 @@
...
 
     @Test
-    void AltBN128MulPrecompiledContract() {
-        final var functionHash = functionEncodeDecoder.functionHashFor("multiplyPoints", EVM_CODES_ABI_PATH);
-        final var serviceParameters = serviceParametersForEvmCodes(functionHash);
-
-        assertThat(contractCallService.processCall(serviceParameters))
-                .isEqualTo(
-                        "0x0769bf9ac56bea3ff40232bcb1b6bd159315d84715b8e679f2d355961915abf02ab799bee0489429554fdb7c8d086475319e63b40b9c5b57cdf1ff3dd9fe2261");
+    void AltBN128MulPrecompiledContract() throws Exception {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java:133]: I know it was that way before, but since we're refactoring we should adjust all methods to start with a lower case letter as per Java standards.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java (hunk 11 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java:35]: Correct. I will open a ticket for that.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java (hunk 12 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java:35]: This will be a separate ticket since I still haven't found a way to avoid using the .bin in the historical tests.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java (hunk 13 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesTest.java]: Comment addressed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java (hunk 11 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java]: Comment addressed

Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallAddressThisTest.java]: This is the new [ticket](https://github.com/hashgraph/hedera-mirror-node/issues/8942) for this

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java (hunk 12 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallEvmCodesHistoricalTest.java:35]: With this corner case we'll need to revisit the other web3j-solidity plugin to generate the bin files from it instead of manually. So I'll reopen this task and work on it https://github.com/hashgraph/hedera-mirror-node/issues/8466.

**Review by steven-sheehy — DISMISSED** [review]
LGTM. Will need to be updated after https://github.com/hashgraph/hedera-mirror-node/pull/8912 is merged to use the new base class

**Review by edwin-greene — DISMISSED** [review]
Looks good

**Comment by kselveliev** [comment]
Updated from main to get changes from #8912

**Review by steven-sheehy — APPROVED** [review]
